### DISCUSSION TITLE
Resize an image on the fly?

### DISCUSSION THREAD

**Discussion by dave-yotta** [SixLabors/ImageSharp#2351] [Q&A] [answered]
Looks like `Image.Load` loads up the whole stream. 

I'm looking for a way to resize a large image (to a much smaller image) keeping a smaller buffer, without any seeking.  

I could probably devise something that has a limited quality by skipping most of the pixels depending on the stride in the pixel format; would rather use a library to do this - is it something I can do with ImageSharp?

**Comment by br3aker** [comment, marked as answer]
If you are using JPEG images - yes, ImageSharp now supports **really** fast downscaled jpeg decoding with scalings 1/2, 1/4 and 1/8 from #2076.

**Reply by antonfirsov** [reply]
Some additional info to the above:

@dave-yotta with 3.0 nightly builds, you should be able to do low-footprint decoding of Jpeg-s thanks to the feature above using the following API:

```C#
DecoderOptions options = new() { TargetSize = new() { Width = 150, Height = 150 } };
using var image = Image.Load(inStream, options);
image.SaveAsJpeg(outStream);
```

Target-sized Load is not available in 2.1, and not optimized for other formats yet in 3.0. Otherwise I believe decoding + resizing is pretty fast, and ImageSharp's pooling memory management helps maintaining a relatively low footprint. Do you have some strongly perf-critical use case, if yes I wonder what?

**Reply by dave-yotta** [reply]
We just have a `GET /api/file/<id>/thumbnail` endpoint (taking dimensions/options), we're able to stream any size through to storage on `POST /api/file` so end up sometimes with several gigabyte images being loaded of any format.  We've also been hit simeltaneously with multiple smaller requests.  It terminates our web api (out of memory) - we're on kubernetes so it recovers but there is still 504s for the other in-flight requests.

We'll likely add a limit to the file size that can be thumbnailed but it would be helpful to avoid that, especially since it might need to be low enough to cope with the unpredictable load patterns.

That's really promising though - does it use a fixed buffer or is it relative to the size or primary dimension of the image being loaded? 

It would be helpful to predict memory usage for the target-sized load to substitute for the file size check against the memory limit - because I don't doubt our users capacity to upload e.g. a 1TB image for some reason.  E.g. `sqrt(fileSize)` or something rough like that?

**Reply by JimBobSquarePants** [reply]
You can use the `Identify` methods to determine the dimensions of the image. Use that in combination with the size of your chosen pixel format and you know the size of the pixel buffer without decoding the image.

**Reply by dave-yotta** [reply]
```csharp
DecoderOptions options = new() { TargetSize = new() { Width = 1024, Height = 100 } };
using var image = Image.Load(inStream, options);
image.SaveAsJpeg(outStream);
```

assuming 1 byte per pixel, do you mean that mean that Image.Load will consume about 100kb memory here for any length of inStream (for jpegs)?  That's pretty good.

Do you have any plans for similar algoritms for the other supported image formats? (is there any issue I can watch?)

Are there any plans to implement stream to stream resizing (or even better system.io.pipelines)? e.g.
```csharp
[Route("api/file/{id}/thumbnail")]
async Task<IActionResult> DownloadThumbnail(string id, [FromQuery] FileDownloadThumbnailWebRequestModel model)
{
    var file = File.Open(id); // example
    HttpContext.Response.RegisterForDispose(file);

    DecoderOptions options = new() { TargetSize = new() { Width = 1024, Height = 100 } };
    var resizeStream = Image.Process(file, new DecodeProcessing(options, model.ResizeMode));
    HttpContext.Response.RegisterForDispose(resizeStream);

    return new StreamResult(resizeStream, $"{id}-thumbnail.jpg", "image/JPEG", true);
}
```

Thanks for the responses - if there's any issues I can watch if you have plans for this stuff It'll help remind me I can remove some of these limits I'm going to have to put in our service, even though they are reasonable limits :).

**Reply by JimBobSquarePants** [reply]
>assuming 1 byte per pixel

Not quite.... A single `Rgb24` pixel would be 3bytes at 1 per color component. `Rgba32` would be 4. We would allocate the buffer for the 1024x100 image and also an internal buffer (_This varies based upon the color/subsampling/cpu but I think the equivalent of no more than a few scanlines_) for the DCT scaling operation within the Jpeg decoder. 

Given the decoder can only decode to 1/2, 1/4, and 1/8 of the original size we may have to allocate an interim buffer to do a final resize to give you the exact dimensions you require. 

We use a sliding window algorithm for that final resize there which limits the memory usage for the operation to 1MB but there would still be the interim buffer allocation for the decoder scaled buffer.

That said, this should lead to really significant memory saving when creating thumbnails compare to v2x. 

Also bear in mind that the memory ImageSharp uses is carefully pooled from an unmanaged memory allocator for all allocations.

As far as other formats go, there are no open issues to track. No other formats we support have this kind of scaling.

**Reply by dave-yotta** [reply]
Ok thanks, just trying to get a handle on the scaling - the buffer within the decoder sounds like it'd scale with sqrt (since it's a few lines) but if it supports a max subsampling of 1/8 then it's more likely that we'll end up using imagesize/8 amount of memory to resize a single image, which could still be problematic for very large images.

**Reply by JimBobSquarePants** [reply]
You can always reject them based upon information procured from `Identify`

**Reply by br3aker** [reply]
> if it supports a max subsampling of 1/8 then it's more likely that we'll end up using imagesize/8 amount of memory to resize a single image

No, result image after initial built-in scaling would be of size `(width  / 8) * (height / 8)` or `width * height / 64`. So it's ~64 times less memory allocated for an image.

Jpeg images' size is contrained to 65535×65535 pixels by the specs, it's physically impossible to save image larger than that using JPEG format.

Decoded black and white (1 byte per pixel) image with size 65535×65535 would allocate 4.294.836.225 bytes or 4 **giga**bytes. This exact image scaled with built-in jpeg scaling would be of size 8192x8192 and would allocate 67.108.864 bytes or 64 **mega**bytes. I assume image processing services not to be bottlenecked by a 500mb RAM hardware :)

Things get a bit trickier with RGB images as they would allocate up to 12 **giga**bytes unscaled and up to 192 **mega**bytes prescaled. IMO this is still acceptable even for personal computers, not even talking about servers.

Regarding internal buffer for actual decoding: worst case scenario won't allocate more than 20 **mega**bytes on top of it regardless if you'd use built-in scaling or not.

**Reply by dave-yotta** [reply]
I can just reject them based on the file size which will be a bit easier - sorry you're right of course, 64 times.  

If we only supported JPEG that would likely be a non-issue, especially given you're using shared buffers.  But we support any image format so we're still at-risk.  The image processing runs with the rest of the APIs, which have 2gb available iirc.

Note: I still prefer the idea of stream processing because it can minimize the response time and buffer sizes.  We've got plans to show these previews in a grid-style layout on clients (including non-images although it doesn't apply here), which worries me within this context - but we'll likely store reduced size images processed in the background in the end so this problem does go away.

**Reply by br3aker** [reply]
Thing is, JPEG format allows us to perform in-place scaling out of the box. Other supported image formats do not have such performance opportunities out of the box, they still can be downscaled inplace - it's just harder to implement, a lot of internal codec code must be rewritten to support this.

Stream processing can help a lot in memory consumption but ImageSharp just doesn't work like that.
Current `Decode -> Image -> Mutate (optional) -> Encode` pipeline just won't work with stream'ed image processing. _Maybe_ one day we'd see such API, maybe not. I'd say an issue can be opened to have an open discussion about this idea with the core team :)

**Reply by JimBobSquarePants** [reply]
The problem with stream processing as I see it is that it makes things super complicated.

We'd have to develop methods to lazily pull the required (different per mutation operation) scanlines from the encoded stream which means we'd either have to copy or lock the stream. You then might have to do some IO thrashing depending on the position of metadata in the stream relative to that data. We'd have to rewrite every single existing operation to handle a sliding window.

Writing the library itself is hard enough as it is with the very limited number of contributors. You'd be better of buying more RAM.

**Reply by dave-yotta** [reply]
Completely understand - it's hard enough to implement one resize operation let alone all the mutations you have here if you have access to (e.g.) only one pixel at a time only once, a limited working buffer and an output stream - becomes a totally different algorithm.
